File: tests/cuda/cpp/test_signal_acoustics.cpp
// tests/cuda/test_signal_acoustics.cpp
#include "insight/insight.h"
#include <cmath>
#include <gtest/gtest.h>
#include <vector>

using namespace ins;

namespace {

class SignalAcousticsTestGPU : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    ins::init({"cpu", "iluvatar"});
    try {
      set_device(GPUPlace(0));
    // Iluvatar: signal API uses F64
    GTEST_SKIP() << "Iluvatar: signal API uses F64 (hardware limit)";
    } catch (...) {
      GTEST_SKIP() << "GPU not available";
    }
  }
};

// ========== mel2hz / hz2mel ==========

TEST_F(SignalAcousticsTestGPU, Mel2HzBasic) {
  // 0 mel = 0 Hz
  std::vector<double> data = {0.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array hz = signal::mel2hz(arr);
  Array hz_cpu = hz.to(CPUPlace());
  EXPECT_NEAR(hz_cpu.data<double>()[0], 0.0, 1e-10);
}

TEST_F(SignalAcousticsTestGPU, Hz2MelBasic) {
  // 0 Hz = 0 mel
  std::vector<double> data = {0.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array mel = signal::hz2mel(arr);
  Array mel_cpu = mel.to(CPUPlace());
  EXPECT_NEAR(mel_cpu.data<double>()[0], 0.0, 1e-10);
}

TEST_F(SignalAcousticsTestGPU, MelHzRoundTrip) {
  // hz2mel(mel2hz(x)) = x
  std::vector<double> data = {100.0, 500.0, 1000.0, 5000.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array mel = signal::hz2mel(arr);
  Array hz_back = signal::mel2hz(mel);
  Array hz_back_cpu = hz_back.to(CPUPlace());
  const double *d = hz_back_cpu.data<double>();
  for (int i = 0; i < 4; ++i) {
    EXPECT_NEAR(d[i], data[i], 1e-6) << "roundtrip failed at " << i;
  }
}

TEST_F(SignalAcousticsTestGPU, Hz2MelMonotonic) {
  std::vector<double> data = {100.0, 200.0, 500.0, 1000.0, 5000.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array mel = signal::hz2mel(arr);
  Array mel_cpu = mel.to(CPUPlace());
  const double *m = mel_cpu.data<double>();
  for (int i = 1; i < 5; ++i) {
    EXPECT_GT(m[i], m[i - 1]) << "not monotonic at " << i;
  }
}

TEST_F(SignalAcousticsTestGPU, MelFrequenciesBasic) {
  Array freqs = signal::mel_frequencies(10, 0.0, 8000.0);
  EXPECT_EQ(freqs.numel(), 10);
  Array freqs_cpu = freqs.to(CPUPlace());
  const double *f = freqs_cpu.data<double>();
  // First should be 0 (or close)
  EXPECT_NEAR(f[0], 0.0, 1.0);
  // Last should be 8000 (or close)
  EXPECT_NEAR(f[9], 8000.0, 100.0);
  // Should be monotonically increasing
  for (int i = 1; i < 10; ++i) {
    EXPECT_GT(f[i], f[i - 1]);
  }
}

TEST_F(SignalAcousticsTestGPU, MelFrequenciesSingle) {
  Array freqs = signal::mel_frequencies(1, 0.0, 8000.0);
  EXPECT_EQ(freqs.numel(), 1);
}

// ========== hz2bark / bark2hz ==========

TEST_F(SignalAcousticsTestGPU, Hz2BarkBasic) {
  // 0 Hz should give bark ~ 0
  std::vector<double> data = {0.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array bark = signal::hz2bark(arr);
  // At hz=0, formula gives 26.81/(1+inf) - 0.53 = -0.53
  // But division by zero issue, so just check it's defined
  EXPECT_TRUE(bark.defined());
}

TEST_F(SignalAcousticsTestGPU, BarkHzRoundTrip) {
  std::vector<double> data = {200.0, 500.0, 1000.0, 4000.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array bark = signal::hz2bark(arr);
  Array hz_back = signal::bark2hz(bark);
  Array hz_back_cpu = hz_back.to(CPUPlace());
  const double *d = hz_back_cpu.data<double>();
  for (int i = 0; i < 4; ++i) {
    EXPECT_NEAR(d[i], data[i], 1.0) << "roundtrip failed at " << i;
  }
}

TEST_F(SignalAcousticsTestGPU, Hz2BarkMonotonic) {
  std::vector<double> data = {100.0, 500.0, 1000.0, 5000.0, 10000.0};
  Array arr = to_array(data).to(GPUPlace(0));
  Array bark = signal::hz2bark(arr);
  Array bark_cpu = bark.to(CPUPlace());
  const double *b = bark_cpu.data<double>();
  for (int i = 1; i < 5; ++i) {
    EXPECT_GT(b[i], b[i - 1]) << "not monotonic at " << i;
  }
}

} // namespace

File: tests/cuda/python/test_signal_acoustics.py
"""Signal acoustics CUDA binding tests."""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr):
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.to(CPU).numpy()


class TestSignalAcousticsCUDA:
    """Acoustic scale conversions — signal_acoustics CUDA tests."""

    def test_mel2hz(self):
        mel = to_gpu(np.array([0.0, 1000.0, 2000.0]))
        hz = ins.signal.mel2hz(mel)
        assert hz is not None
        assert hz.numel == 3

    def test_hz2mel(self):
        hz = to_gpu(np.array([0.0, 1000.0, 4000.0]))
        mel = ins.signal.hz2mel(hz)
        assert mel is not None
        assert mel.numel == 3

    def test_mel_roundtrip(self):
        hz_orig = np.array([100.0, 500.0, 1000.0, 4000.0, 8000.0])
        mel = ins.signal.hz2mel(to_gpu(hz_orig))
        hz_back = ins.signal.mel2hz(mel)
        np.testing.assert_allclose(to_numpy(hz_back), hz_orig, rtol=1e-5)

    def test_mel_frequencies(self):
        freqs = ins.signal.mel_frequencies(128, fmin=0.0, fmax=11025.0)
        freqs_gpu = freqs.to(GPU)
        assert freqs_gpu is not None
        assert freqs_gpu.numel == 128

    def test_mel_frequencies_custom(self):
        freqs = ins.signal.mel_frequencies(64, fmin=20.0, fmax=8000.0)
        freqs_gpu = freqs.to(GPU)
        assert freqs_gpu is not None
        assert freqs_gpu.numel == 64

    def test_hz2bark(self):
        hz = to_gpu(np.array([100.0, 500.0, 2000.0]))
        bark = ins.signal.hz2bark(hz)
        assert bark is not None
        assert bark.numel == 3

    def test_bark2hz(self):
        bark = to_gpu(np.array([1.0, 5.0, 15.0]))
        hz = ins.signal.bark2hz(bark)
        assert hz is not None
        assert hz.numel == 3

    def test_bark_roundtrip(self):
        hz_orig = np.array([100.0, 500.0, 2000.0, 6000.0])
        bark = ins.signal.hz2bark(to_gpu(hz_orig))
        hz_back = ins.signal.bark2hz(bark)
        np.testing.assert_allclose(to_numpy(hz_back), hz_orig, rtol=1e-3)

    def test_mel_frequencies_sorted(self):
        freqs = ins.signal.mel_frequencies(128, fmin=0.0, fmax=11025.0)
        freqs_np = freqs.to(CPU).numpy()
        assert np.all(np.diff(freqs_np) >= 0)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
